### PR TITLE
Bypass the "Checks" page and go directly to the Travis build when clicking the "Details" links on a PR

### LINKED ISSUE #1620: Add a shortcut to Travis build (bypass Check tab)
Example of pull requests with Travis PR/Branch Checks: https://github.com/Kocal/semantic-release-preset/pull/7 and https://github.com/Kocal/semantic-release-preset/pull/4.

#### The problem

Before GitHub Checks, we had the possibility to directly go on Travis website by clicking on `details` here:

selection_999 195

Now we should make 3 or 4 clicks and it's really boring:

selection_999 194

selection_999 196

selection_999 197

#### Suggestion

The big issue here is to retrieve Travis build URL. I don't know if it's possible to do through GitHub v3/4 API, but we can do:
- For each Travis Checks:
  - Extract `a.status-actions[href]` value
  - Make a XHR request
  - Parse the response body and find the `<a>` containg `View more details on Travis CI`, and extract `href` value

I have three suggestion at the moment. 

#### Visual update

I have three ideas on how to exploit Travis url.

Completely replace `Details` link, easy and clean:
selection_999 202

Make `Travis CI - ...` a link, less easy, would break CSS:
selection_999 199

Or create a new link and put it aside `Details` link, hard and horrible to see, would break some CSS too:
selection_999 200

Thanks

### PR DESCRIPTION
**Pull request by Kocal**
Will close #1620.

It was too hard to use API v3/v4 to get last checks from a pull request, so I used the easy solution:
  - find Travis branch/PR checks
  - extract `Details` link
  - Make an XHR request
  - Find `View more ...` link

And this is working very well:

capture d ecran du 2018-11-24 22-27-59

### CODE REVIEW

File: source/features/bypass-checks-travis.js [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+	if (response.ok) {
+		const dom = domify(await response.text());
+		const links = select.all('[id^="check_run_"] a', dom);
+		const viewMoreLink = links.find(isViewMoreLink);
+
+		if (viewMoreLink) {
+			details.setAttribute('href', viewMoreLink.getAttribute('href'));
+			details.setAttribute('rel', viewMoreLink.getAttribute('rel'));
+			details.setAttribute('target', viewMoreLink.getAttribute('target'));
```
Comment (bfred-it) [inline source/features/bypass-checks-travis.js]: All of these are properties as well, I think, so you can use `details.rel = viewMoreLink.rel`

Comment (bfred-it) [reply source/features/bypass-checks-travis.js]: Also is this the behavior we want? Should it _replace_ the previous link?

Comment (Kocal) [reply source/features/bypass-checks-travis.js]: I always prefered to access/set attributes with `getAttribute`/`setAttribute` , but here I think we can directly use properties yes.

For the behavior, this was my first suggestion in #1620.
I think it's not a problem to totally replace the link `Details` , no one is actually crazy enough to see Travis checks in `Checks` tab.

Comment (Kocal) [reply source/features/bypass-checks-travis.js]: > The suggestion to use props stands

Updated.

**Comment by Kocal** [comment]
Ready for next review

**Comment by Kocal** [comment]
Your changes are interesting, less lines = less bugs 😛 

But I'm not a big fan about using styling classes as selectors (e.g.: `[href^="https://travis-ci.com"].text-small`)

**Comment by bfred-it** [comment]
It's more efficient than loop+`textContent` and no less brittle (both rely on GitHub not changing what we expect) so between the two I prefer the class.

**Comment by Kocal** [comment]
Should I update the README myself or you will do it?

**Comment by Kocal** [comment]
Updated

File: readme.md [hunk 2]
```diff
@@ -179,6 +179,7 @@ And [many more…](source/content.css)
 - [Filter PRs that you haven't submitted a review for.](https://user-images.githubusercontent.com/9264728/42569205-0c32a376-8510-11e8-9406-b3ddcf6087ef.png)
 - Use <kbd>←</kbd> and <kbd>→</kbd> to navigate through pages with pagination.
 - [Find a user’s most starred repositories in their profile.](https://user-images.githubusercontent.com/1402241/48474026-43e3ae80-e82c-11e8-93de-159ad4c6f283.png)
+- [Bypass the new Checks page to directly see your Travis builds](https://user-images.githubusercontent.com/2103975/49071220-c6596e80-f22d-11e8-8a1e-bdcd62aa6ece.png)
```
Comment (sindresorhus) [inline readme.md]: I think I would be more specific here.

```suggestion
- [Bypass the "Checks" page and go directly to the Travis build when clicking the "Details" links on a PR](https://user-images.githubusercontent.com/2103975/49071220-c6596e80-f22d-11e8-8a1e-bdcd62aa6ece.png)
```

@bfred-it ^?

File: source/features/bypass-checks-travis.js [hunk 3]
```diff
@@ -0,0 +1,25 @@
+import select from 'select-dom';
+import domify from '../libs/domify';
+
+export default async function () {
+	// If anything errors, RGH will display the error next to the feature name
+	await Promise.all(select.all('[href="/apps/travis-ci"]').map(bypass));
```
[the remark this answers is not in the document]
Comment (sindresorhus) [reply source/features/bypass-checks-travis.js]: Not clear what you're asking thoughts on. This seems like the best way to implement this regardless of error reporting.

File: source/features/bypass-checks-travis.js [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+}
+
+async function bypass(check) {
+	const details = select('.status-actions', check.parentNode);
+
+	const response = await fetch(details.href, {
+		credentials: 'include'
+	});
+
+	if (response.ok) {
```
Comment (sindresorhus) [inline source/features/bypass-checks-travis.js]: Nitpick, but would be nicer with an early return here instead.

File: source/features/bypass-checks-travis.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (bfred-it) [reply source/features/bypass-checks-travis.js]: If we don't care about the logs we could just use a regular loop like I previously implemented this, without essentially awaiting each `bypass` call:

https://github.com/sindresorhus/refined-github/commit/c3f769d36df3bf466e7f28de0342c4788c984946

File: readme.md (hunk 2 above)
Comment (Kocal) [reply readme.md]: That's definitely sounds better

**Comment by Kocal** [comment]
https://github.com/sindresorhus/refined-github/pull/1636#pullrequestreview-179650980 and https://github.com/sindresorhus/refined-github/pull/1636#pullrequestreview-179651250 applied, thanks for the review.

File: source/features/bypass-checks-travis.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (sindresorhus) [reply source/features/bypass-checks-travis.js]: Ah, I see. I cannot wait for better async stack traces so we don't have to write code specifically for better debugging.

**Comment by Kocal** [comment]
I guess this [error on Travis](https://travis-ci.org/sindresorhus/refined-github/builds/461154704#L497) is unrelated:
selection_305

Did something on master happened? :stuck_out_tongue_winking_eye:

**Comment by Kocal** [comment]
Thanks, I think this is mergeable now I guess

**Comment by sindresorhus** [comment]
Great work, @Kocal :)

**Comment by Kocal** [comment]
Happy to help :)

**Pull request merged**